src/libraries/System.IO.FileSystem/tests/FileStream/DevicesPipesAndSockets.cs

```diff
@@ -0,0 +1,253 @@
+                    await ReadByteAsync(fs, 42);
+                }),
+                Task.Run(async () => 
+                {
+                    using var fs = File.OpenWrite(fifoPath);
+                    await WriteByteAsync(fs, 42);
+                }));
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.AnyUnix & ~TestPlatforms.Browser)]
+        public unsafe void SocketPair_ReadWrite()
+        {
+            const int AF_UNIX = 1;
+            const int SOCK_STREAM = 1;
+            int* ptr = stackalloc int[2];
+            Assert.Equal(0, socketpair(AF_UNIX, SOCK_STREAM, 0, ptr));
+
+            using var readFileStream = new FileStream(new SafeFileHandle((IntPtr)ptr[0], ownsHandle: false), FileAccess.Read);
+            using var writeFileStream = new FileStream(new SafeFileHandle((IntPtr)ptr[1], ownsHandle: false), FileAccess.Write);
+
+            Task.WhenAll(
+                Task.Run(() => ReadByte(readFileStream, 42)),
+                Task.Run(() => WriteByte(writeFileStream, 42))).GetAwaiter().GetResult();
+
+            close(ptr[0]);
+            close(ptr[1]);
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.AnyUnix & ~TestPlatforms.Browser)]
+        public void SocketPair_ReadWrite_Async()
+        {
+            const int AF_UNIX = 1;
+            const int SOCK_STREAM = 1;
+            unsafe
+            {
+                int* ptr = stackalloc int[2];
+                Assert.Equal(0, socketpair(AF_UNIX, SOCK_STREAM, 0, ptr));
+
+                using var readFileStream = new FileStream(new SafeFileHandle((IntPtr)ptr[0], ownsHandle: false), FileAccess.Read);
+                using var writeFileStream = new FileStream(new SafeFileHandle((IntPtr)ptr[1], ownsHandle: false), FileAccess.Write);
+
+                Task.WhenAll(
+                    ReadByteAsync(readFileStream, 42),
+                    WriteByteAsync(writeFileStream, 42)).GetAwaiter().GetResult();
+
+                close(ptr[0]);
+                close(ptr[1]);
+            }
+        }
+
+        private static void ReadByte(FileStream fs, byte expected)
+        {
+            var buffer = new byte[1];
+            Assert.Equal(1, fs.Read(buffer));
+            Assert.Equal(expected, buffer[0]);
+        }
+
+        private static void WriteByte(FileStream fs, byte value)
+        {
+            fs.Write(new byte[] { value });
+            fs.Flush();
+        }
+
+        private static async Task ReadByteAsync(FileStream fs, byte expected)
+        {
+            var buffer = new byte[1];
+            Assert.Equal(1, await fs.ReadAsync(buffer));
+            Assert.Equal(expected, buffer[0]);
+        }
+
+        private static async Task WriteByteAsync(FileStream fs, byte value)
+        {
+            await fs.WriteAsync(new byte[] { value });
+            await fs.FlushAsync();
+        }
+
+        private static bool IsDeviceUnreachable(string devicePath, FileStreamOptions? options)
+        {
```

Comment:
Is this for `/dev/tty`, or also for other files?

If it's only for `/dev/tty`, can we remove  `/dev/tty` from `DevicePaths` and get rid of `IsDeviceUnreachable`? or is `/dev/tty` special in some way for this test?

Reply:
I don't think removing /dev/tty test makes sense when this PR is fixing the very issue. :)

POSIX requires three paths - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap10.html:
```
Contains /dev/console, /dev/null, and /dev/tty, described below.
```

A portable code would assert that these are available and only assume that console and zero are optional.

Reply:
I don't like methods like `IsDeviceUnreachable` which cause tests to be silently skipped.

If `/dev/tty` is needed for coverage, I'm fine keeping it.

Reply:
I don't think /dev/tty needs additional check. My code example in the issue was overly defensive. If _anything_, it should be /dev/console /dev/zero requiring a simple File.Exists() check to simply skip the test.